src/plugins/intel_gna/frontend/quantization.cpp

```diff
@@ -200,262 +54,228 @@ float ScaleFactorForQuantization(void *ptr_float_memory, float target_max, size_
+                THROW_GNA_EXCEPTION << "invalid channel multiplier: " << channel_multiplier;
             }
 
-            // range checking
-            ptr_int_weights[offset] = static_cast<int8_t>(normalizedWeight);
+            // channel multiplier shouldn't be 0
+            channel_multiplier = channel_multiplier == 0 ? 1 : channel_multiplier;
+            multiplier = static_cast<uint8_t> (channel_multiplier + 0.5f);
         }
 
-        for (uint32_t j = num_columns; j < num_columns_padded; j++) {
-            ptr_int_weights[i * num_columns + j] = 0;
-        }
-    }
+        for (uint32_t col = 0; col < data.num_columns; col++) {
+            auto offset = row * data.num_columns + col;
+            float rounding_value = (ptr_float_weights[offset] > 0) ? 0.5f : -0.5f;
+            float value = ptr_float_weights[offset];
 
-    for (uint32_t i = num_rows; i < num_rows_padded; i++) {
-        for (uint32_t j = 0; j < num_columns_padded; j++) {
-            ptr_int_weights[i * num_columns + j] = 0;
-        }
-        ptr_int_biases[i].multiplier = 0;
-    }
+            if (data.fq_num_stats > 0) {
+                auto x = value;
+                if (x <= std::min(input_low, input_high)) {
+                    value = output_low;
+                } else if (x > std::max(input_low, input_high)) {
+                    value = output_high;
+                } else {
+                    value = nearbyint((x - input_low) / (input_high - input_low) * (levels - 1)) / (levels - 1) *
+                                (output_high - output_low) + output_low;
+                }
+            }
 
-    if (ptr_float_biases != nullptr) {
-        for (uint32_t j = 0; j < num_rows; j++) {
-            float rounding_value = (ptr_float_biases[j] > 0) ? 0.5f : -0.5f;
-            float value = ptr_float_biases[j] * *ptr_output_scale_factor + rounding_value;
-            if (value > 2147483647.0) {
-                ptr_int_biases[j].bias = 2147483647L;
+            if (rich) {
+                if (quantizedWeights) {
+                    value -= MAX_VAL_1B_WEIGHT;
+                } else {
+                    value = value * (ptr_weight_scale_factor / multiplier) + rounding_value;
+                }
+            } else {
+                value = value * ptr_weight_scale_factor + rounding_value;
+            }
+
+            int8_t* ptr_weight_8 = ptr_int_weights + offset;
+
+            if (value > std::numeric_limits<int8_t>::max()) {
+                *ptr_weight_8 = std::numeric_limits<int8_t>::max();
                 num_saturate++;
-            } else if (value < -2147483648.0) {
-                ptr_int_biases[j].bias = -2147483648LL;
+            } else if (value < std::numeric_limits<int8_t>::min()) {
+                *ptr_weight_8 = std::numeric_limits<int8_t>::min();
                 num_saturate++;
             } else {
-                ptr_int_biases[j].bias = (int32_t) value;
+                *ptr_weight_8 = (int8_t)value;
```

Comment:
Use static_cast here?

Reply:
Thanks for all the comments, my PR is still WIP. Anyway, I'll take them into account.